Converting Local --> Sync (#1238)
mongodb/docs-realm: source/sync/local-to-sync.txt

   :depth: 2
   :class: singlecol

Overview
--------
If you have a mobile app that uses a local {+client-database+} and you want to 
convert the app to use {+sync+}, there are a few steps you 
need to follow. You first need to set up a {+backend-short+} 
backend, and then you make code changes in your client app(s). 

In the following sections, you will learn to do the following on the backend:

- Create a new {+realm+} app in {+atlas+},
- Configure at least one User Authentication provider, and
- Choose a partition strategy.

And then, in your client app, you will:

- Update your {+service-short+} object models,
- Connect to the new synced {+realm+} with a sync configuration file,
- Authenticate the user.

Finally, if needed, you will also need to copy your local data to the new 
synced {+realm+}.

Create a Backend App
--------------------
To sync your data between devices and with {+atlas+}, you first need to create 
a {+app+}. This app serves -- at a minimum -- two purposes: it provides a 
gateway to the data, and it provides data security in the form of user 
authentication, data encryption, and access control.

To get started creating the backend app, follow the steps in 

> :ref:`Create a Realm App (Realm UI) <create-a-realm-app>`.

Reviewer: ```suggestion
:ref:`Create a {+app+} (Realm UI) <create-a-realm-app>`.
```

of your choice.

Partition the Data 
~~~~~~~~~~~~~~~~~~
Data :ref:`partitioning <partitioning>` strategies enable you to configure which 
users have access to which data. This process is often the most time-consuming 
part of setting up {+sync+}, as you as you consider what partition strategy 
is the best fit for your use case and data.

As part of determining your :ref:`partition strategy <partition-strategies>`, 
you'll choose a :ref:`partition key <partition-keys>`. You might choose a 
partition key that is already a part of your object models, which you can then 
specify when you set up {+sync+}. If you have chosen a partition key that 
isn't already a part of your object models, you'll need to update those models.

Update the Client App
---------------------
Now that you have the backend set up for {+sync-short+}, you need to make a few 
changes in your client app. Follow the steps in the Quick Start for your 
language/platform:

- :ref:`<android-client-quick-start-sync>`
- :ref:`<ios-client-quick-start-with-sync>`
- :ref:`<dotnet-client-quick-start>`
- :ref:`<react-native-client-quick-start>`
- :ref:`<node-client-quick-start>`

Copy Existing Data
----------------------

> None of the data in your local realm will be automatically transferred to your 

Reviewer: ```suggestion
None of the data in your local {+realm+} will be automatically transferred to your 
```

> new synced realm; it is up to you to ensure the data you want to preserve is 

Reviewer: ```suggestion
new synced {+realm+}; it is up to you to ensure the data you want to preserve is 
```

> copied to the new realm. How you go about doing this depends, in part, on your 

Reviewer: ```suggestion
copied to the new {+realm+}. How you go about doing this depends, in part, on your 
```

specific needs, but the following diagram and steps outline the process for 

> copying all of the data from the local realm to the synced realm. If done 

Reviewer: ```suggestion
copying all of the data from the local {+realm+} to the synced {+realm+}. If done 
```

correctly, the impact on app performance after the initial copy will be 
minimal.

App Logic
~~~~~~~~~

.. figure:: /images/local_to_sync.png
   :alt: Flow chart showing the steps to convert from local to sync

> 1. Check if a synced realm already exists.

Reviewer: ```suggestion
1. Check if a synced {+realm+} already exists.
```

> 2. If not, create one. Open a connection to it and to the local realm.

Reviewer: ```suggestion
2. If not, create one. Open a connection to it and to the local {+realm+}.
```

> 3. Read each record from the local realm and modify it to meet the needs of the 

Reviewer: ```suggestion
3. Read each record from the local {+realm+} and modify it to meet the needs of the 
```

   synced realm (adding a partition key, for example).

> 4. Copy the modified record to the new synced realm. Sync will automatically 

Reviewer: ```suggestion
4. Copy the modified record to the new synced {+realm+}. Sync will automatically 
```

   copy the record to Atlas.

> 5. Confirm that all records you want to preserve are in the new realm.

Reviewer: ```suggestion
5. Confirm that all records you want to preserve are in the new {+realm+}.
```

> 6. Delete the local realm file.

Reviewer: ```suggestion
6. Delete the local {+realm+} file.
```

> 7. On each subsequent app load, simply check if the local realm has been deleted.

Reviewer: ```suggestion
7. On each subsequent app load, simply check if the local {+realm+} has been deleted.
```

.. _realm-sync-local-to-sync:

=======================================

> Convert a Local Realm to a Synced Realm

Reviewer: I wonder if we should instead name this "Add Sync to an Application that Already Uses Realm Database"... or something like that

=======================================

.. default-domain:: mongodb

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

Overview
--------
If you have a mobile app that uses a local {+client-database+} and you want to 
convert the app to use {+sync+}, there are a few steps you 
need to follow. You first need to set up a {+backend-short+} 
backend, and then you make code changes in your client app(s). 

In the following sections, you will learn to do the following on the backend:

- Create a new {+realm+} app in {+atlas+},

> - Configure a User Authentication provider, and

Reviewer: ```suggestion
- Configure at least one User Authentication provider, and
```

To sync your data between devices and with {+atlas+}, you first need to create 
a {+realm+} app. This app serves -- at a minimum -- two purposes: it provides a 
gateway to the data, and it provides data security in the form of user 
authentication, data encryption, and access control.

To get started, follow the steps in 
:ref:`Create a Realm App (Realm UI) <create-a-realm-app>`.

Enable Authentication
~~~~~~~~~~~~~~~~~~~~~
User authentication plays a crucial role in {+sync+}. {+realm+} provides 
many different authentication providers; once you have configured one or more,
you can then control exactly who has access to which data. From a Sync perspective, 
it doesn't matter which authentication method you choose.

To get started, learn about the 
:ref:`Authentication Providers <authentication-providers>` and then 
follow the configuration steps for the provider(s) of your choice.

Partition the Data 
~~~~~~~~~~~~~~~~~~
Data :ref:`partitioning <partitioning>` strategies enable you to configure which 
users have access to which data. This process is often the most time-consuming 
part of setting up {+sync+}, as you want to consider the ramifications of 
different strategies.

> Once you determine your 

Author: Much better.

:ref:`partition strategy <partition-strategies>` and choose a
:ref:`partition key <partition-keys>`, you may need to update your existing object 
models to include the partition field, unless you have chosen an existing field.

Update the Client App
---------------------
Now that you have the backend set up for Sync, you need to make a few changes in 

> your client app. Since each SDk is different, choose the tab below for your 

Reviewer: ```suggestion
your client app. Since each SDK is different, choose the tab below for your 
```

platform / language to get started:

.. tabs-realm-sdks::

   .. tab::
      :tabid: android

>       1. To use {+backend+} features such as authentication and sync, you must

Reviewer: With the Android SDK, before a user can use any sync functionality (such as the `App` class), they have to [enable sync](https://docs.mongodb.com/realm/sdk/android/install/#add-realm-to-the-application-level-build-gradle-file) in their gradle config. (described in the linked step, but i'm breaking that out into a linkable example in a PR soon)

         access your {+app+} using your {+app+} ID. You can find your {+app+} ID 
         in the {+ui+}.

         .. tabs-realm-languages::

            .. tab::
               :tabid: java

               .. literalinclude:: /examples/generated/android/sync/MainActivity.codeblock.initialize-the-app.java
                  :language: java
            
            .. tab::
               :tabid: kotlin

               .. literalinclude:: /examples/generated/android/sync/MainActivity.codeblock.initialize-the-app.kt
                  :language: kotlin

      2. You have already defined one or more object models. If these models 
         do not currently have a property that maps to the partition key you 

>          selected when creating the backend, you'll need to add that property. 

Reviewer: The partition key isn't actually necessary in the Android SDK, so we can remove this step entirely.

            
            .. tab::
               :tabid: kotlin

               .. literalinclude:: /examples/generated/android/sync/MainActivity.codeblock.define-object-model.kt
                  :language: kotlin

      3. Authenticate the user by calling the appropriate authentication method, 
         based on which provider you have set up. The following example uses 
         Anonymous authentication; for details on implementing each of the 
         providers, see :ref:`android-authenticate`.

         .. tabs-realm-languages::

            .. tab::
               :tabid: java

               .. literalinclude:: /examples/generated/android/sync/MainActivity.codeblock.authenticate-a-user.java
                  :language: java
                  :emphasize-lines: 1-3
            
            .. tab::
               :tabid: kotlin

               .. literalinclude:: /examples/generated/android/sync/MainActivity.codeblock.authenticate-a-user.kt
                  :language: kotlin
                  :emphasize-lines: 1-3

      4. You can now open a synced {+realm+}. Use a 
         :java-sdk:`SyncConfiguration <io/realm/mongodb/sync/SyncConfiguration.html>`

>          object to specify of how your application synchronizes data with 

Reviewer: ```suggestion
         object to control how your application synchronizes data with 
```

>          {+backend+}. The following example includes a partition name and the user 

Reviewer: ```suggestion
         {+backend+}. The following example includes a partition and the user 
```
It might be better to use either "partition value" or just "partition" here so we don't introduce a new term in a step guide.

                  name: 'string',
                  status: 'string',
               },
               primaryKey: '_id',
            };
            }

            export {Task};

      4. Authenticate the user by calling the appropriate authentication method, 
         based on which provider you have set up. The following example uses 
         Anonymous authentication; for details on implementing each of the 
         providers, see :ref:`react-native-authenticate-users`.

         .. code-block:: javascript
            :emphasize-lines: 8,9

            import Realm from 'realm';         

            async function anonymousLogin() {
               let user;
               try {
                  // pass in the appConfig variable that you created earlier
                  const app = new Realm.App(appConfig); 
      
                  // create an anonymous credential
                  const credentials = Realm.Credentials.anonymous(); 
                  user = await app.logIn(credentials);
                  return user;

               } catch (error) {
                     throw `Error logging in anonymously: ${JSON.stringify(error,null,2)}`;
               }
            }

>       5. Once you have :ref:`enabled {+sync+} <enable-sync>` and authenticated a

Reviewer 2: I liked how you called out specifically what was different about the opening a synced Realm code examples in the other SDKs (i.e. passing in a config/partition/user object). I was bummed not to see it in the RN or Node.js examples. Perhaps we could add that? I think it helps the reader focus on what's different here vs. in the standard "open a Realm" example.
